panic: runtime error: index out of range in recheckFile

```
14:11:25 INFO: syncthing v1.1.4 "Erbium Earthworm" (go1.12.5 linux-arm) teamcity@build.syncthing.net 2019-05-12 19:17:55 UTC
Panic at 2019-06-26T04:32:03+02:00
panic: runtime error: index out of range

goroutine 456147 [running]:
github.com/syncthing/syncthing/lib/model.(*model).recheckFile(0x1424180, 0xce349c4c, 0x8c35d347, 0xb2a7fe8, 0x6a548915, 0xbe69fb4d, 0x73efe85a, 0x1e47cbc7, 0x65d409ed, 0x9f1cf0, ...)
	/opt/tcagent/syncthing-1-work/174e136266f8a219/lib/model/model.go:1597 +0xed0
github.com/syncthing/syncthing/lib/model.(*model).Request(0x1424180, 0xce349c4c, 0x8c35d347, 0xb2a7fe8, 0x6a548915, 0xbe69fb4d, 0x73efe85a, 0x1e47cbc7, 0x65d409ed, 0x22c29c0, ...)
	/opt/tcagent/syncthing-1-work/174e136266f8a219/lib/model/model.go:1571 +0x15cc
github.com/syncthing/syncthing/lib/protocol.(*rawConnection).handleRequest(0x1eccf00, 0x4d, 0x22c29c0, 0xb, 0x22c29e0, 0xe, 0xab800000, 0x1, 0x400000, 0x22dc8a0, ...)
	/opt/tcagent/syncthing-1-work/174e136266f8a219/lib/protocol/protocol.go:602 +0xd0
created by github.com/syncthing/syncthing/lib/protocol.(*rawConnection).readerLoop
	/opt/tcagent/syncthing-1-work/174e136266f8a219/lib/protocol/protocol.go:409 +0x52c
```

https://github.com/syncthing/syncthing/blob/e4956358fbab0d78e975d474acba2a0d098e4adc/lib/model/model.go#L1591-L1597

The `int(offset)` expression can overflow on 32 bit archs, leading to a negative block index.

There is a similar but less serious issue here:

https://github.com/syncthing/syncthing/blob/e4956358fbab0d78e975d474acba2a0d098e4adc/lib/model/model.go#L703

--- a/lib/model/folder_sendrecv.go
+++ b/lib/model/folder_sendrecv.go
@@ -1256,7 +1256,7 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 			if len(hashesToFind) > 0 {
 				file, err = f.fs.Open(state.file.Name)
 				if err == nil {
-					weakHashFinder, err = weakhash.NewFinder(f.ctx, file, int(state.file.BlockSize()), hashesToFind)
+					weakHashFinder, err = weakhash.NewFinder(f.ctx, file, state.file.BlockSize(), hashesToFind)
 					if err != nil {
 						l.Debugln("weak hasher", err)
 					}

--- a/lib/model/model.go
+++ b/lib/model/model.go
@@ -722,7 +722,7 @@ func (m *model) Completion(device protocol.DeviceID, folder string) FolderComple
 		}
 
 		// This might might be more than it really is, because some blocks can be of a smaller size.
-		downloaded = int64(counts[ft.Name] * int(ft.BlockSize()))
+		downloaded = int64(counts[ft.Name]) * int64(ft.BlockSize())
 
 		fileNeed = ft.FileSize() - downloaded
 		if fileNeed < 0 {
@@ -1623,7 +1623,7 @@ func (m *model) recheckFile(deviceID protocol.DeviceID, folderFs fs.Filesystem, 
 		return
 	}
 
-	blockIndex := int(offset) / cf.BlockSize()
+	blockIndex := int(offset / int64(cf.BlockSize()))
 	if blockIndex >= len(cf.Blocks) {
 		l.Debugf("%v recheckFile: %s: %q / %q i=%d: block index too far", m, deviceID, folder, name, blockIndex)
 		return

